compiler/rustc_hir_analysis/src/check/intrinsicck.rs

```diff
@@ -451,10 +451,26 @@ impl<'a, 'tcx> InlineAsmCtxt<'a, 'tcx> {
                         );
                     }
                 }
-                // No special checking is needed for these:
-                // - Typeck has checked that Const operands are integers.
-                // - AST lowering guarantees that SymStatic points to a static.
-                hir::InlineAsmOperand::Const { .. } | hir::InlineAsmOperand::SymStatic { .. } => {}
+                hir::InlineAsmOperand::Const { anon_const } => {
```

Comment:
a suggestion: you can move this check into `fn anon_const_type_of` and return the returned type with `ty::Error` if it's not an int. THis should also fix the ICE in mir borrowck

Reply:
I tried this here https://github.com/rust-lang/rust/compare/master...folkertdev:rust:const-asm-type-fix and that makes the test work! However, error messages are now emitted out of order. I'm not sure how big a problem that is, and how to fix it if needed.

Reply:
as an extra note, this issue appears to be entirely unrelated to https://github.com/rust-lang/rust/issues/96304, lifetimes work fine in these const expressions.

```rust
core::arch::global_asm!("/* {} */", const Foo::<'static>::X);

struct Foo<'a>(&'a ());

impl<'a> Foo<'a> {
    const X: u64 = 42;
}
```

Reply:
:+1: yeah, that's my preferred solution, though I'd maybe keep the check in `intrinsicck` as an assert in case `type_of` gets changed in the future. I don't think the message order is something we should worry about here.

Lifetimes in the expression work, it's lifetime in the type of the const that cause issues, i.e. changing the type of `X` to `&'static str` for example